device/sycl/src/finding/find_tracks.hpp

```diff
@@ -0,0 +1,502 @@
+
+    // Number of tracks per seed
+    vecmem::data::vector_buffer<unsigned int> n_tracks_per_seed_buffer(n_seeds,
+                                                                       mr.main);
+    copy.setup(n_tracks_per_seed_buffer)->wait();
+
+    // Create a map for links
+    std::map<unsigned int, vecmem::data::vector_buffer<candidate_link>>
+        link_map;
+
+    // Create a buffer of tip links
+    vecmem::data::vector_buffer<typename candidate_link::link_index_type>
+        tips_buffer{config.max_num_branches_per_seed * n_seeds, mr.main,
+                    vecmem::data::buffer_type::resizable};
+    copy.setup(tips_buffer)->wait();
+
+    // Link size
+    std::vector<std::size_t> n_candidates_per_step;
+    n_candidates_per_step.reserve(config.max_track_candidates_per_track);
+
+    unsigned int n_in_params = n_seeds;
+    for (unsigned int step = 0;
+         step < config.max_track_candidates_per_track && n_in_params > 0;
+         step++) {
+
+        /*****************************************************************
+         * Kernel2: Apply material interaction
+         ****************************************************************/
+
+        queue
+            .submit([&](::sycl::handler& h) {
+                h.parallel_for<
+                    typename kernels_t::apply_interaction_kernel_type>(
+                    calculate1DimNdRange(n_in_params, 64),
+                    [config, det, n_in_params,
+                     in_params = vecmem::get_data(in_params_buffer),
+                     param_liveness = vecmem::get_data(param_liveness_buffer)](
+                        ::sycl::nd_item<1> item) {
+                        device::apply_interaction<
+                            typename navigator_t::detector_type>(
+                            details::global_index(item), config,
+                            {det, n_in_params, in_params, param_liveness});
+                    });
+            })
+            .wait_and_throw();
+
+        /*****************************************************************
+         * Kernel3: Find valid tracks
+         *****************************************************************/
+
+        // Previous step
+        const unsigned int prev_step = (step == 0 ? 0 : step - 1);
+
+        // Buffer for kalman-updated parameters spawned by the
+        // measurement candidates
+        const unsigned int n_max_candidates =
+            n_in_params * config.max_num_branches_per_surface;
+
+        bound_track_parameters_collection_types::buffer updated_params_buffer(
+            n_in_params * config.max_num_branches_per_surface, mr.main);
+        copy.setup(updated_params_buffer)->wait();
+
+        vecmem::data::vector_buffer<unsigned int> updated_liveness_buffer(
+            n_in_params * config.max_num_branches_per_surface, mr.main);
+        copy.setup(updated_liveness_buffer)->wait();
+        copy.memset(updated_liveness_buffer, 0)->wait();
```

Comment:
Why is this necessary?

Reply:
As step 1 I was just not caring about efficiency. Only about correctness.

Indeed, there are 2 "more correct" ways of doing this:
  - If we require nothing about the queue that is given to the algorithm, then we need to start holding on to the "event objects" returned by these calls, and only wait for them in the appropriate places.
    * Though this will not give too much of an improvement unfortunately.
  - We can require `::sycl::queue` objects to always be "in order". (https://github.khronos.org/SYCL_Reference/iface/queue.html#is-in-order) Then we can implement this function practically in the same way between CUDA and SYCL.

It's the latter that I'll want to go for. But decided not to make this "first" PR even more complicated with that. :thinking:

Reply:
I guess this is a reply to https://github.com/acts-project/traccc/pull/811#discussion_r1908679856, right? :thinking:

Reply:
No but wait I am talking about the memset operation itself. Why is that necessary? I don't remember that being in the CUDA code.

Reply:
My understanding is that allocated memory is not guaranteed to be null-initialized with SYCL. I'm happy to be proven wrong though.

Reply:
It's not a point about automatic zeroing. Rather, if I recall correctly (which I may not), the `find_tracks` kernel guarantees that the updated liveness is written to a meaningful value before the array is read for all indices below `n_candidates`: https://github.com/acts-project/traccc/blob/main/device/common/include/traccc/finding/device/impl/find_tracks.ipp; if that's true then the memset is unnecessary.

Have you found that the code doesn't run without this memset?

Reply:
I just now looked at it. This buffer is being set in the `find_tracks` kernel, with the common function only ever setting elements of the buffer to `1`.
  - https://github.com/acts-project/traccc/blob/main/device/common/include/traccc/finding/device/impl/find_tracks.ipp#L230
  - https://github.com/acts-project/traccc/blob/main/device/common/include/traccc/finding/device/impl/find_tracks.ipp#L286

So we'll either have to teach that kernel to set `0` values as well, or teach the CUDA code to do this memset.

Reply:
:confused: What's the point of the buffer then? If the first N elements of the track container are guaranteed to be "live", then what do we use these liveness buffers for at all? I assumed that tracks would be allowed to be "sparsely alive". :confused:

But sure, if we can kill these buffers, I'm all for it. :thinking:

Reply:
The unit tests succeed without it as well, so let's remove it for now.

But we should have a discussion about the logic-flow here soon. Apparently it could be simplified a bit.